machinelearning/llama.cpp.md

```diff
@@ -0,0 +1,73 @@
+
+# How to use llama.cpp on Graviton CPUs
+
+Building from sources is the recommended way to use llama.cpp on Graviton CPUs, and for other hardware platforms too. This section provides the instructions on how to build it from sources.
+
+**Prerequisites**
+
+It is highly recommended to use the AMIs based on Linux Kernel 5.10 and beyond for the best LLM inference performance on Graviton3 instances. The following queries can be used to list down the AMIs with Kernel 5.10 and beyond.
+
+```
+# For Kernel 5.10 based AMIs list
+aws ec2 describe-images --owners amazon --filters "Name=architecture,Values=arm64" "Name=name,Values=*kernel-5.10*" --query 'sort_by(Images, &CreationDate)[].Name'
+
+# For Kernel 6.x based AMIs list
+aws ec2 describe-images --owners amazon --filters "Name=architecture,Values=arm64" "Name=name,Values=*kernel-6.*" --query 'sort_by(Images, &CreationDate)[].Name'
+```
+
+**Build llama.cpp from source**
+
+```
+# Clone llama.cpp
+git clone https://github.com/ggerganov/llama.cpp
+cd llama.cpp
+
+# build with make
+make
+
+```
+
+# Run LLM inference with llama.cpp
+
+llama.cpp provides a set of tools to (1) convert model binary file into GPT-Generated Unified Format (GGUF) (2) quantize single and half precision format models into one of the quantized formats, for example into Q4_0 (3) rearrange model weights into blocked layout to leverage the hardware specific GEMM kernels and improve the performance and (4) run LLM inference locally.
+
+The following instructions use Meta Llama-3 8B paramater model from [Hugging Face](https://huggingface.co/models) models repository to demonstrate LLM inference performance on AWS Graviton based EC2 instances. The model is already availble in GGUF and quantized formats. For the steps on how to convert model binary into GGUF format, please check [llama.cpp README](https://github.com/ggerganov/llama.cpp/blob/master/README.md) document. The model from the Hugging Face model repository can be run directly on Graviton based EC2 instances. However, to be able to leverage the optimized GEMM kernels and improve the LLM inference performance by up to 2x, it is recommended to rearrange the model weights into blocked layout which is a one time model pre-processing step. This section shows you how to rearrange the model weights into blocked layout and run it efficiently on AWS Graviton3 and Graviton4 based Amazon EC2 instances.
```

Comment:
EC2 instances needs to be capitalized in order to keep with our branding, just FYI.

Reply:
please check the README . (drop document)